airflow/api_connexion/endpoints/task_instance_endpoint.py

```diff
@@ -14,39 +14,243 @@
+        query = query.filter(or_(*[key == v for v in values]))
+    return query
+
+
+def _apply_range_filter(query, key, value_range: Tuple[Any, Any]):
+    gte_value, lte_value = value_range
+    if gte_value is not None:
+        query = query.filter(key >= gte_value)
+    if lte_value is not None:
+        query = query.filter(key <= lte_value)
+    return query
 
 
+@format_parameters(
+    {
+        "execution_date_gte": format_datetime,
+        "execution_date_lte": format_datetime,
+        "start_date_gte": format_datetime,
+        "start_date_lte": format_datetime,
+        "end_date_gte": format_datetime,
+        "end_date_lte": format_datetime,
+    }
+)
 @security.requires_authentication
-def get_task_instances():
+@provide_session
+def get_task_instances(
+    limit: int,
+    dag_id: Optional[str] = None,
+    dag_run_id: Optional[str] = None,
+    execution_date_gte: Optional[str] = None,
+    execution_date_lte: Optional[str] = None,
+    start_date_gte: Optional[str] = None,
+    start_date_lte: Optional[str] = None,
+    end_date_gte: Optional[str] = None,
+    end_date_lte: Optional[str] = None,
+    duration_gte: Optional[float] = None,
+    duration_lte: Optional[float] = None,
+    state: Optional[str] = None,
+    pool: Optional[List[str]] = None,
+    queue: Optional[List[str]] = None,
+    offset: Optional[int] = None,
+    session=None,
+):  # pylint: disable=too-many-arguments
     """
-    Get list of task instances of DAG.
+    Get list of task instances.
     """
-    raise NotImplementedError("Not implemented yet.")
+    query = session.query(TI)
+
+    if dag_id != "~":
+        query = query.filter(TI.dag_id == dag_id)
+    if dag_run_id != "~":
+        query = query.join(DR, and_(TI.dag_id == DR.dag_id, TI.execution_date == DR.execution_date))
+        query = query.filter(DR.run_id == dag_run_id)
+    query = _apply_range_filter(
+        query,
+        key=DR.execution_date,
+        value_range=(execution_date_gte, execution_date_lte),
+    )
+    query = _apply_range_filter(query, key=TI.start_date, value_range=(start_date_gte, start_date_lte))
+    query = _apply_range_filter(query, key=TI.end_date, value_range=(end_date_gte, end_date_lte))
+    query = _apply_range_filter(query, key=TI.duration, value_range=(duration_gte, duration_lte))
+    query = _apply_array_filter(query, key=TI.state, values=state)
+    query = _apply_array_filter(query, key=TI.pool, values=pool)
+    query = _apply_array_filter(query, key=TI.queue, values=queue)
+
+    query = query.join(
+        SlaMiss,
+        and_(
+            SlaMiss.dag_id == TI.dag_id,
+            SlaMiss.task_id == TI.task_id,
+            SlaMiss.execution_date == TI.execution_date,
+        ),
+        isouter=True,
+    ).add_entity(SlaMiss)
+    task_instances = query.offset(offset).limit(limit).all()
+    total_entries = session.query(
+        func.count(func.distinct(TI.dag_id, TI.execution_date, TI.task_id))
+    ).scalar()
```

Comment:
@mik-laj Passing multiple columns to distinct works perfectly with postgres but raises error in case of sqlite how should we handle this case.

Reply:
It seems to me that the count query doesn't need join, so we can simplify everything and improve SQL compatibility.